src/wp-includes/class-wp-block-supports.php

```diff
@@ -209,3 +211,37 @@ function get_block_wrapper_attributes( $extra_attributes = array() ) {
 	return implode( ' ', $normalized_attributes );
 }
 
+
+/**
+ * Callback hooked to the register_block_type_args filter.
+ *
+ * This hooks into block registration to wrap the render_callback
+ * of dynamic blocks with a closure that keeps track of the
+ * current block to be rendered.
+ *
+ * @param array $args Block attributes.
+ * @return array Block attributes.
+ */
+function wp_block_supports_track_block_to_render( $args ) {
```

Comment:
I think in core we shouldn't be extending the render_callback like that we should just inject the variable assignment in the render function.